Compile warning in Climate component [-Wclass-memaccess]

### The problem

When building an esphome project using the climate component, the output contains a warning about clearing an object in a wrong way. This is done here: https://github.com/esphome/esphome/blob/953f0569fbf060cc066a9f05b702824a04f5c2d1/esphome/components/climate/climate.cpp#L334

`warning: 'void* memset(void*, int, size_t)' clearing an object of non-trivial type 'struct esphome::climate::ClimateDeviceRestoreState'; use assignment or value-initialization instead [-Wclass-memaccess]`

While this is not a bug per se, it does stand out in the build log in HA as it is shown in yellow, which made me think something might be broken, or break in the near future.

The full compile warning message:
```
Compiling /data/irco/.pioenvs/irco/src/esphome/components/climate_ir/climate_ir.cpp.o
src/esphome/components/climate/climate.cpp: In member function 'void esphome::climate::Climate::save_state_()':
src/esphome/components/climate/climate.cpp:334:54: warning: 'void* memset(void*, int, size_t)' clearing an object of non-trivial type 'struct esphome::climate::ClimateDeviceRestoreState'; use assignment or value-initialization instead [-Wclass-memaccess]
  334 |   memset(&state, 0, sizeof(ClimateDeviceRestoreState));
      |                                                      ^
In file included from src/esphome/components/climate/climate.cpp:1:
src/esphome/components/climate/climate.h:125:8: note: 'struct esphome::climate::ClimateDeviceRestoreState' declared here
  125 | struct ClimateDeviceRestoreState {
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~
```

### Which version of ESPHome has the issue?

2022.2.1

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2022.2.6

### What platform are you using?

ESP8266

### Board

board: esp12e; framework: arduino; platform: platformio/espressif8266 @ 3.2.0

### Component causing the issue

climate

### Example YAML snippet

_No response_

### Anything in the logs that might be useful for us?

```txt
Compiling /data/irco/.pioenvs/irco/src/esphome/components/climate_ir/climate_ir.cpp.o
src/esphome/components/climate/climate.cpp: In member function 'void esphome::climate::Climate::save_state_()':
src/esphome/components/climate/climate.cpp:334:54: warning: 'void* memset(void*, int, size_t)' clearing an object of non-trivial type 'struct esphome::climate::ClimateDeviceRestoreState'; use assignment or value-initialization instead [-Wclass-memaccess]
  334 |   memset(&state, 0, sizeof(ClimateDeviceRestoreState));
      |                                                      ^
In file included from src/esphome/components/climate/climate.cpp:1:
src/esphome/components/climate/climate.h:125:8: note: 'struct esphome::climate::ClimateDeviceRestoreState' declared here
  125 | struct ClimateDeviceRestoreState {
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~
```


### Additional information

There seems to be an attempt to block this warning in certain cases already in the code on [L330](https://github.com/esphome/esphome/blame/dev/esphome/components/climate/climate.cpp#L330), but it doesn't seem to trigger for this board type and/or build platform?

--- a/esphome/components/climate/climate.cpp
+++ b/esphome/components/climate/climate.cpp
@@ -1,4 +1,5 @@
 #include "climate.h"
+#include "esphome/core/macros.h"
 
 namespace esphome {
 namespace climate {
@@ -326,14 +327,16 @@ optional<ClimateDeviceRestoreState> Climate::restore_state_() {
   return recovered;
 }
 void Climate::save_state_() {
-#if defined(USE_ESP_IDF) && !defined(CLANG_TIDY)
+#if (defined(USE_ESP_IDF) || (defined(USE_ESP8266) && USE_ARDUINO_VERSION_CODE >= VERSION_CODE(3, 0, 0))) && !defined(CLANG_TIDY)
 #pragma GCC diagnostic ignored "-Wclass-memaccess"
+#define TEMP_IGNORE_MEMACCESS
 #endif
   ClimateDeviceRestoreState state{};
   // initialize as zero to prevent random data on stack triggering erase
   memset(&state, 0, sizeof(ClimateDeviceRestoreState));
-#if USE_ESP_IDF && !defined(CLANG_TIDY)
+#if TEMP_IGNORE_MEMACCESS
 #pragma GCC diagnostic pop
+#unset TEMP_IGNORE_MEMACCESS
 #endif
 
   state.mode = this->mode;

--- a/esphome/components/climate/climate.cpp
+++ b/esphome/components/climate/climate.cpp
@@ -334,7 +334,7 @@ void Climate::save_state_() {
   ClimateDeviceRestoreState state{};
   // initialize as zero to prevent random data on stack triggering erase
   memset(&state, 0, sizeof(ClimateDeviceRestoreState));
-#if TEMP_IGNORE_MEMACCESS
+#ifdef TEMP_IGNORE_MEMACCESS
 #pragma GCC diagnostic pop
 #undef TEMP_IGNORE_MEMACCESS
 #endif

--- a/esphome/components/climate/climate.cpp
+++ b/esphome/components/climate/climate.cpp
@@ -327,7 +327,8 @@ optional<ClimateDeviceRestoreState> Climate::restore_state_() {
   return recovered;
 }
 void Climate::save_state_() {
-#if (defined(USE_ESP_IDF) || (defined(USE_ESP8266) && USE_ARDUINO_VERSION_CODE >= VERSION_CODE(3, 0, 0))) && !defined(CLANG_TIDY)
+#if (defined(USE_ESP_IDF) || (defined(USE_ESP8266) && USE_ARDUINO_VERSION_CODE >= VERSION_CODE(3, 0, 0))) && \
+    !defined(CLANG_TIDY)
 #pragma GCC diagnostic ignored "-Wclass-memaccess"
 #define TEMP_IGNORE_MEMACCESS
 #endif

--- a/esphome/components/climate/climate.cpp
+++ b/esphome/components/climate/climate.cpp
@@ -336,7 +336,7 @@ void Climate::save_state_() {
   memset(&state, 0, sizeof(ClimateDeviceRestoreState));
 #if TEMP_IGNORE_MEMACCESS
 #pragma GCC diagnostic pop
-#unset TEMP_IGNORE_MEMACCESS
+#undef TEMP_IGNORE_MEMACCESS
 #endif
 
   state.mode = this->mode;

